File: packages/jexlang-java/src/main/java/com/jexlang/java/scopes/Scope.java
package com.jexlang.java.scopes;

import com.jexlang.java.eval.errors.JexLangRuntimeError;
import com.jexlang.java.types.JexNull;
import com.jexlang.java.types.JexValue;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

public class Scope {

    public enum ScopeType {
        GLOBAL,
        PROGRAM,
        BLOCK
    }

    private final Scope parentScope;

    private final Map<String, JexValue> variables;
    private final Set<String> constants;
    private final ScopeType scopeType;

    public Scope(
            Scope parentScope,
            ScopeType scopeType
    ) {
        this.parentScope = parentScope;
        this.scopeType = Objects.requireNonNullElse(scopeType, ScopeType.PROGRAM);
        this.variables = new HashMap<>();
        this.constants = new java.util.HashSet<>();
    }

    public Scope getParentScope() {
        return parentScope;
    }

    public void declareVariable(String name, JexValue value, boolean isConst) {
        if (this.variables.containsKey(name)) {
            throw new JexLangRuntimeError("Variable '" + name + "' is already declared.");
        }
        this.variables.put(name, value);
        if (isConst) {
            this.constants.add(name);
        }
    }

    public void assignVariable(String name, JexValue value) {
        Scope scope = this.resolveScope(name);
        if (scope == null) {
            throw new JexLangRuntimeError("Variable '" + name + "' is not declared.");
        }
        if (scope.constants.contains(name)) {
            throw new JexLangRuntimeError("Variable '" + name + "' is a constant and cannot be re-declared or modified.");
        }
        scope.variables.put(name, value);
    }

    public Scope resolveScope(String name) {
        if (this.variables.containsKey(name)) {
            return this;
        } else if (this.parentScope != null) {
            return this.parentScope.resolveScope(name);
        } else {
            return null;
        }
    }

    public Scope resolveScope(ScopeType scopeType) {
        if (this.scopeType == scopeType) {
            return this;
        } else if (this.parentScope != null) {
            return this.parentScope.resolveScope(scopeType);
        } else {
            return null;
        }
    }

    public JexValue getVariable(String name) {
        if (this.variables.containsKey(name)) {
            return this.variables.get(name);
        }

        // If not in this scope, check parent scope.
        if (this.parentScope != null) {
            return this.parentScope.getVariable(name);
        }

        return new JexNull();
    }

    public boolean hasVariable(String name) {
        return this.resolveScope(name) != null;
    }
}

File: packages/jexlang-swift/jexlang-swift/eval/scopes/Scope.swift
//
//  Scope.swift
//  jexlang-swift
//
//  Created by Vinay Kumar on 04/11/25.
//

import Foundation

public final class Scope {

    // MARK: - ScopeType Enum
    public enum ScopeType {
        case global
        case program
        case block
    }

    // MARK: - Properties
    private let parentScope: Scope?
    private var variables: [String: JexValue]
    private var constants: Set<String>
    private let scopeType: ScopeType

    // MARK: - Initializer
    public init(parentScope: Scope? = nil, scopeType: ScopeType) {
        self.parentScope = parentScope
        self.scopeType = scopeType
        self.variables = [:]
        self.constants = []
    }

    // MARK: - Scope Access
    public func getParentScope() -> Scope? {
        return parentScope
    }

    // MARK: - Variable Declaration
    public func declareVariable(_ name: String, value: JexValue, isConst: Bool) {
        let isGlobalScope = (self.scopeType == .global)

        if variables[name] != nil && !isGlobalScope {
            NSException.raise(jexLangError: JexLangRuntimeError(message: "Variable '\(name)' is already declared."))
            return
        }

        variables[name] = value

        if isConst {
            constants.insert(name)
        } else if !isConst && constants.contains(name) && isGlobalScope {
            // Delete from constants if re-declared as non-const for the same variable in global scope
            constants.remove(name)
        }
    }

    // MARK: - Variable Assignment
    public func assignVariable(_ name: String, value: JexValue) {
        guard let scope = resolveScope(for: name) else {
            NSException.raise(jexLangError: JexLangRuntimeError(message: "Variable '\(name)' is not declared."))
            return
        }

        if scope.constants.contains(name) {
            NSException.raise(jexLangError: JexLangRuntimeError(message: "Variable '\(name)' is a constant and cannot be re-declared or modified."))
            return
        }

        scope.variables[name] = value
    }

    // MARK: - Declare or Assign
    public func declareAndAssignVariable(_ name: String, value: JexValue, isConst: Bool) {
        if variables[name] != nil {
            assignVariable(name, value: value)
        } else {
            declareVariable(name, value: value, isConst: isConst)
        }
    }

    // MARK: - Scope Resolution
    public func resolveScope(for name: String) -> Scope? {
        if variables[name] != nil {
            return self
        } else if let parent = parentScope {
            return parent.resolveScope(for: name)
        } else {
            return nil
        }
    }

    public func resolveScope(ofType type: ScopeType) -> Scope? {
        if self.scopeType == type {
            return self
        } else if let parent = parentScope {
            return parent.resolveScope(ofType: type)
        } else {
            return nil
        }
    }

    // MARK: - Accessors
    public func getAllVariables() -> [String: JexValue] {
        return variables
    }

    public func getVariable(_ name: String) -> JexValue {
        if let val = variables[name] {
            return val
        }

        if let parent = parentScope {
            return parent.getVariable(name)
        }

        return JexNull()
    }

    public func hasVariable(_ name: String) -> Bool {
        return resolveScope(for: name) != nil
    }
}

